torch/_inductor/fx_passes/post_grad.py

```diff
@@ -479,6 +482,50 @@ def is_valid_splitwithsizes_cat(match):
     return True
 
 
+
+def reinplace_scatters(graph):
+    """
+    Reinplaces scatter operations in easy cases where the node being mutated
+    is only used by the scatter (users == 1)
+
+    Also handles input mutations when there is a corresponding copy node.
```

Comment:
It's not obvious to me why we expect this to give sound results. In particular you don't have alias tracking so you don't know if it may not be safe to reinplace because there is another tensor that aliases you that would be improperly updated by this change

Reply:
I think you could add alias tracking by looking at `meta["val"].untyped_storage()` easily enough. alternatively you could grab `fw_metadata` / `bw_metadata`, from the Tracing Context, and follow the input to `index_put` while its input is a view, until you either get to an input of the graph, or you find that it is produced by a node which is not a view.

Reply:
yea, using the storage from fake vals is sufficient

Reply:
This is only OK if Inductor doesn't introduce new aliasing, e.g., because it converted a clone() into a no-op on the fly. So no, it's not clear to me this is sufficient either.

Reply:
> This is only OK if Inductor doesn't introduce new aliasing, e.g., because it converted a clone() into a no-op on the fly.

In principle, Inductor should "support" in-place operators. So this has always been possible, but if it actually happens then it would be a bug.

Any decompositions would be applied during trace-time and so their fake tensors should be correct.

Reply:
We may need to update pattern existing matcher passes to use this new propagation for this to be sound.

Reply:
I'd like to do this in a follow-up PR. This PR has already expanded quite a bit, and I don't think the current state is any worse than the existing one.